pkg/statistics/handle/autoanalyze/autoanalyze.go

```diff
@@ -76,6 +79,167 @@ func (sa *statsAnalyze) DeleteAnalyzeJobs(updateTime time.Time) error {
+func (sa *statsAnalyze) CleanupCorruptedAnalyzeJobsOnDeadInstances() error {
+	return statsutil.CallWithSCtx(sa.statsHandle.SPool(), func(sctx sessionctx.Context) error {
+		return CleanupCorruptedAnalyzeJobsOnDeadInstances(sctx)
+	})
+}
+
+// SelectAnalyzeJobsOnCurrentInstanceSQL is the SQL to select the analyze jobs whose
+// state is `pending` or `running` and the update time is more than 10 minutes ago
+// and the instance is current instance.
+const SelectAnalyzeJobsOnCurrentInstanceSQL = `SELECT id, process_id
+		FROM mysql.analyze_jobs
+		WHERE instance = %?
+		AND state IN ('pending', 'running')
+		AND TIMESTAMPDIFF(MINUTE, update_time, NOW()) > 10`
+
+// SelectAnalyzeJobsSQL is the SQL to select the analyze jobs whose
+// state is `pending` or `running` and the update time is more than 10 minutes ago.
+const SelectAnalyzeJobsSQL = `SELECT id, instance
+		FROM mysql.analyze_jobs
+		WHERE state IN ('pending', 'running')
+		AND TIMESTAMPDIFF(MINUTE, update_time, NOW()) > 10`
+
+// BatchUpdateAnalyzeJobSQL is the SQL to update the analyze jobs to `failed` state.
+const BatchUpdateAnalyzeJobSQL = `UPDATE mysql.analyze_jobs
+            SET state = 'failed',
+            fail_reason = 'TiDB Server is down when running the analyze job',
+            process_id = NULL
+            WHERE id IN (%?)`
+
+// CleanupCorruptedAnalyzeJobsOnCurrentInstance cleans up the potentially corrupted analyze job from current instance.
+// Exported for testing.
+func CleanupCorruptedAnalyzeJobsOnCurrentInstance(
+	sctx sessionctx.Context,
+	currentRunningProcessIDs map[uint64]struct{},
+) error {
+	serverInfo, err := infosync.GetServerInfo()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	instance := net.JoinHostPort(serverInfo.IP, strconv.Itoa(int(serverInfo.Port)))
+	// Get all the analyze jobs whose state is `pending` or `running` and the update time is more than 10 minutes ago
+	// and the instance is current instance.
+	rows, _, err := statsutil.ExecRows(
+		sctx,
+		SelectAnalyzeJobsOnCurrentInstanceSQL,
+		instance,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	jobIDs := make([]uint64, 0, len(rows))
+	for _, row := range rows {
+		// The process ID is typically non-null for running or pending jobs.
+		// However, in rare cases(I don't which case), it may be null. Therefore, it's necessary to check its value.
+		if !row.IsNull(1) {
+			processID := row.GetUint64(1)
+			// If the process id is not in currentRunningProcessIDs, we need to clean up the job.
+			// They don't belong to current instance any more.
+			if _, ok := currentRunningProcessIDs[processID]; !ok {
+				jobID := row.GetUint64(0)
+				jobIDs = append(jobIDs, jobID)
+			}
+		}
+	}
+
+	// Do a batch update to clean up the jobs.
+	if len(jobIDs) > 0 {
+		// Jobs ID to string list.
+		var jobIDStrs []string
+		for _, jobID := range jobIDs {
+			jobIDStrs = append(jobIDStrs, strconv.FormatUint(jobID, 10))
+		}
```

Comment:
Do we have a better way to do this? It seems the framework can not help us to do it.

Reply:
Oh I didn't need it